### PR TITLE
MonadIO instead of IO and chipmunk function addPostStepCallback

### PR DESCRIPTION
**Pull request by Simre1**
I replaced all apecs-physics instances to work within MonadIO and not just IO. Additionally, I added the function addPostStepCallback since I needed it for my project.

### CODE REVIEW

File: apecs-physics/src/Apecs/Physics/Shape.hs [hunk 1]
```diff
@@ -245,26 +246,26 @@ setFriction shape (realToFrac -> friction) = [C.exp| void {
...
+instance MonadIO m => ExplGet m (Space Friction) where
+  explExists s ety = liftIO $ explExists (cast s :: Space Shape) ety
+  explGet (Space _ sMap _ _ _) ety = liftIO $ do
     Just (Record s _) <- M.lookup ety <$> readIORef sMap
     Friction <$> getFriction s
 
 -- SurfaceVelocity
 getSurfaceVelocity :: Ptr Shape -> IO Vec
-getSurfaceVelocity shape = do
+getSurfaceVelocity shape = liftIO $ do
```
Comment (jonascarpay) [inline apecs-physics/src/Apecs/Physics/Shape.hs]: This `liftIO` is unnecessary, right?

File: apecs-physics/src/Apecs/Physics/Shape.hs [hunk 2]
```diff
@@ -277,26 +278,26 @@ setSurfaceVelocity shape (V2 (realToFrac -> x) (realToFrac -> y)) = [C.block| vo
...
+instance MonadIO m => ExplGet m (Space SurfaceVelocity) where
+  explExists s ety = liftIO $ explExists (cast s :: Space Shape) ety
+  explGet (Space _ sMap _ _ _) ety = liftIO $ do
     Just (Record s _) <- M.lookup ety <$> readIORef sMap
     SurfaceVelocity <$> getSurfaceVelocity s
 
 -- CollisionFilter
 getFilter :: Ptr Shape -> IO CollisionFilter
-getFilter shape = do
+getFilter shape = liftIO $ do
```
Comment (jonascarpay) [inline apecs-physics/src/Apecs/Physics/Shape.hs]: Same here

File: apecs-physics/src/Apecs/Physics/Collision.hs [hunk 3]
```diff
@@ -41,7 +43,7 @@ defaultHandler :: CollisionHandler
 defaultHandler = CollisionHandler (Wildcard 0) Nothing Nothing Nothing Nothing
 
 mkCollision :: Ptr Collision -> IO Collision
-mkCollision arb = do
+mkCollision arb = liftIO $ do
```
Comment (jonascarpay) [inline apecs-physics/src/Apecs/Physics/Collision.hs]: And here

File: apecs-physics/src/Apecs/Physics/Collision.hs [hunk 4]
```diff
@@ -117,26 +120,39 @@ destroyCollisionHandler = error "Destroy CollisionHandler not yet implemented"
...
-instance ExplGet IO (Space CollisionHandler) where
-  explExists (Space _ _ _ hMap _) ety = M.member ety <$> readIORef hMap
-  explGet (Space _ _ _ hMap _) ety = do
+instance (MonadIO m) => ExplGet m (Space CollisionHandler) where
+  explExists (Space _ _ _ hMap _) ety = liftIO $ M.member ety <$> readIORef hMap
+  explGet (Space _ _ _ hMap _) ety = liftIO $ do
     Just (Record _ handler) <- M.lookup ety <$> readIORef hMap
     return handler
 
+addPostStepCallback :: (Has w m Physics, MonadIO m) => Int -> SystemT w IO () -> SystemT w m ()
```
Comment (jonascarpay) [inline apecs-physics/src/Apecs/Physics/Collision.hs]: Would you mind adding some haddocks on how to use this? Or maybe just provide a link to https://chipmunk-physics.net/release/ChipmunkLatest-Docs/#CollisionCallbacks-PostStep

**Comment by Simre1** [comment]
I hope I got it all right, since I don't have a compiler at hand right now.

**Comment by jonascarpay** [comment]
Thanks! CI says it's good, let's merge

**Pull request merged**